### PR TITLE
code refactor to reduce code duplicate and avoid manual toBatch operation.

### PR DESCRIPTION
**Pull request by hhbyyh**
## What changes were proposed in this pull request?

1. Support more data types for features and label columns.
2. Move DLParam from SparkVersion to dl folder, to avoid code duplicate between Spark 1.6 and 2.0+
3. add transient to model parameter in DLEstimator and DLClassifier, to avoid unnecessary serialization.
4. use Optimizer and Predictor for batch training and avoid manual toBatch operation.

Other new functions will be added in follow-up PRs.

## How was this patch tested?

existing unit tests

### CODE REVIEW

**Comment by hhbyyh** [comment]
cc @wzhongyuan @yiheng @jason-dai

File: spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala [hunk 1]
```diff
@@ -67,68 +166,61 @@ class DLEstimator[@specialized(Float, Double) T: ClassTag](
...
-   * learning rate for the optimizer in the DLEstimator.
-   * Default: 0.001
-   */
-  val learningRate = new DoubleParam(this, "learningRate", "learningRate", ParamValidators.gt(0))
-
-  setDefault(learningRate -> 1e-3)
-
-
-  def getLearningRate: Double = $(learningRate)
+  setDefault(maxEpoch -> 100)
```
Comment (jason-dai) [inline spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: 100 epoch can be too large?

File: spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala [hunk 2]
```diff
@@ -209,52 +276,40 @@ class DLModel[@specialized(Float, Double) T: ClassTag](
...
-    val resultRDD = dataset.rdd.zip(predictRdd).map { case (row, predict) =>
-      Row.fromSeq(row.toSeq ++ Seq(predict))
-    }
-    val resultSchema = transformSchema(dataset.schema)
-    dataset.sqlContext.createDataFrame(resultRDD, resultSchema)
+    val prediction = model.asInstanceOf[Module[T]]
+      .predict(samplesRDD, $(batchSize))
+      .map { t => outputToPrediction(t.asInstanceOf[Tensor[T]]) }
+
+    val resultRDD = dataFrame.rdd.zipWithIndex().map(_.swap)
```
Comment (jason-dai) [inline spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: A potential issue is that dataFrame.rdd may need to be computed twice if it's not cached. One way to address that is to do something like https://github.com/intel-analytics/BigDL/blob/master/spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala#L93

File: spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala (hunk 1 above)
Comment (hhbyyh) [reply spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: Just want to ensure that we got converged model on some smaller data set. 
I can change it to 50.

**Comment by hhbyyh** [comment]
Please let me know if further change is required. Thanks.

**Comment by hhbyyh** [comment]
http://172.168.2.101:8080/job/BigDL-PR-V/568/

**Comment by hhbyyh** [comment]
@jason-dai we need the change to support Mastercard, is it OK to merge this into 0.4?

File: spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala [hunk 3]
```diff
@@ -281,25 +283,31 @@ class DLModel[@specialized(Float, Double) T: ClassTag](
     val featureType = dataFrame.schema($(featuresCol)).dataType
     val featureColIndex = dataFrame.schema.fieldIndex($(featuresCol))
     val featureFunc = getConvertFunc(featureType)
+    val sc = dataFrame.sqlContext.sparkContext
+    val modelBroadCast = ModelBroadcast[T]().broadcast(sc, model)
+    val localBatchSize = $(batchSize)
+
+    val resultRDD = dataFrame.rdd.mapPartitions { rowIter =>
+      val rows = rowIter.toArray
```
Comment (jason-dai) [inline spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: What if rows do not fit in memory?

Comment (hhbyyh) [reply spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: Updated to use group iterator

**Comment by jason-dai** [comment]
LGTM

**Comment by hhbyyh** [comment]
http://172.168.2.101:8080/job/BigDL-PR-V/584/ passed.

**Pull request merged**